feat: Add role api keys in UI

novakzaballa opened the pull request:
Thanks for submitting a PR! Please check the boxes below:

- [x] I have run [`pre-commit`](https://docs.flagsmith.com/platform/contributing#pre-commit) to check linting
- [x] I have filled in the "Changes" section below?
- [x] I have filled in the "How did you test this code" section below?
- [x] I have used a [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) title for this Pull Request

## Changes

- Add the ability in the frontend to create role API keys

## How did you test this code?

- Go to manage -> Keys
- Click the "Create API Key" button
- Disable the "Is admin" switch and add roles

matthewelwell commented on api/api_keys/models.py:
```diff
@@ -18,3 +20,13 @@ class MasterAPIKey(AbstractAPIKey, SoftDeleteObject):
     objects = MasterAPIKeyManager()
     is_admin = models.BooleanField(default=True)
+
+    if settings.IS_RBAC_INSTALLED:
```
This should be inside the method itself.

novakzaballa replied:
Moved.

matthewelwell commented on api/api_keys/serializers.py:
```diff
@@ -11,17 +14,21 @@ class MasterAPIKeySerializer(serializers.ModelSerializer):
         "for every endpoint apart from create",
     )
     is_admin = serializers.BooleanField(default=True)
+    if settings.IS_RBAC_INSTALLED:
+        roles = MasterAPIKeyRoleSerializer(many=True, read_only=True, source="role")
```
Hmm... I don't really like this behaviour of changing serializers based on the existence of external libraries. Can we think of another way to do this? Perhaps, as per the suggestion on the other PR, by creating the view in the rbac repo and adding it to URLs in the core repo. We do this for a number of views in the flagsmith-workflows repo for example.

novakzaballa replied:
I saw that you had a similar comment in another PR, yes I think you are right, the serializers would no longer depend on RBAC and could be tested more easily

matthewelwell replied:
So can we remove it from here now?

matthewelwell commented on api/api_keys/serializers.py:
```diff
@@ -36,3 +43,21 @@ def validate_is_admin(self, is_admin: bool):
+    def to_representation(self, instance):
+        if self.context["view"].action == "retrieve":
+            is_admin = instance.is_admin if hasattr(instance, "is_admin") else False
+            if is_admin is False and not settings.IS_RBAC_INSTALLED:
+                raise serializers.ValidationError(
+                    "RBAC is not installed, cannot create non-admin key"
+                )
```
This can just be done in a `validate_is_admin` method on the serializer.

matthewelwell replied:
Also, I'm confused why we're validating this on retrieve, shouldn't this only need to be validated on create?

novakzaballa replied:
Good catch @matthewelwell, this code is unnecessary, deleted.

I also found a bug, it was already solved, I had to create a PR for the RBAC: https://github.com/Flagsmith/flagsmith-rbac/pull/19

kyle-ssg reviewed (approved):
FE looks fine from a code perspective 👍

matthewelwell commented on api/api_keys/serializers.py:
```diff
@@ -11,17 +14,21 @@ class MasterAPIKeySerializer(serializers.ModelSerializer):
             "expiry_date",
             "key",
             "is_admin",
+            *([] if not settings.IS_RBAC_INSTALLED else ["roles"]),
```
This should be removed too I think?

novakzaballa replied:
Both were removed.

matthewelwell commented:
@novakzaballa I've just tested this but it seems as though the roles aren't getting applied to the keys correctly. 

https://github.com/Flagsmith/flagsmith/assets/14089968/82ca8c8a-c75e-4aef-8235-279d7299382e

matthewelwell reviewed (commented):
Code looks ok, but as per previous comment, it doesn't appear to be working. Mostly leaving this review comment to remove this PR from my list because GH doesn't seem to have a function to do it another way.

novakzaballa commented:
> @novakzaballa I've just tested this but it seems as though the roles aren't getting applied to the keys correctly.
> 

@matthewelwell I found the issue, and now is solved

matthewelwell commented:
Yep, seems like it works now which is great! A couple of minor things: 

1. This shouldn't say 'Terraform API Keys' any more.  
<img width="1258" alt="Screenshot 2024-02-15 at 11 45 16" src="https://github.com/Flagsmith/flagsmith/assets/14089968/7517f3c8-37d4-4e4f-b065-4b271d21dda9">

2. Can we add an item to the table indicating whether a key is an admin or not?
3. We should probably add an information dialog stating that "Terraform API Keys have become API Keys which can be used for all purposes"

novakzaballa commented:
> Yep, seems like it works now which is great! A couple of minor things:
> 
> 1. This shouldn't say 'Terraform API Keys' any more.
> 
> 2. Can we add an item to the table indicating whether a key is an admin or not?
> 3. We need to add / update the docs. Currently they're linking to the Terraform documentation which is not correct. Perhaps we should create a separate PR for the docs?


Hey @matthewelwell, I changed the title and added a field in the API keys table to see if it is an admin API key or not.

I will create an issue for the API keys documentation.
